File: src/org/likexin/twopointer/QuickSort.java
package org.likexin.twopointer;

public class QuickSort {

  /**
   * 先整体有序再部分有序。即：先找准一个基准数 pivot，将小于 pivot 的数扔到左边，大于 pivot 的数扔到右边。这样一遍下来已经算是宏观意义上的有序，再递归到左右两边使左右两边都有序。
   * 时间复杂度：O(nlogn)
   * 空间复杂度：O(1)
   */
  public void sortInteger(int[] A) {
    if (A == null || A.length == 0) {
      return;
    }

    quickSort(A, 0, A.length - 1);
  }

  private void quickSort(int[] A, int start, int end) {
    if (start >= end) {
      return;
    }

    int left = start, right = end;
    // 1. pivot is value not index
    int pivot = A[(start + end) / 2];

    // 2. left <= right not left < right,left < right will make stack overflow error
    while (left <= right) {
      // 3. A[left] < pivot not A[left] <= pivot
      while (left <= right && A[left] < pivot) {
        left++;
      }
      while (left <= right && A[right] > pivot) {
        right--;
      }

      if (left <= right) {
        int temp = A[left];
        A[left] = A[right];
        A[right] = temp;
      }
    }

    quickSort(A, left, end);
    quickSort(A, start, right);
  }

}

File: JavaScript/sort/QuickSort.js
/**
 * 平均时间复杂度：O(nlongn)，最坏情况：已经排好序的数组。
 * 
 * 过程：找到一个基准点，将小于该基准点的数放到基准点左边，大于基准点的数放到基准点右边。再对左右两边数组递归找基准点并比较的过程。
 * @param {arr} arr 
 */
function quickSort(arr) {
  if (arr.length < 2) {
    return arr;
  }
  return quick(arr, 0, arr.length - 1);
}

function quick(arr, left, right) {
  let index = partition(arr, left, right);
  if (left < index - 1) {
    quick(arr, left, index - 1);
  }
  if (index < right) {
    quick(arr, index, right);
  }
  return arr;
}

function partition(arr, left, right) {
  let pivot = arr[Math.floor((right + left) / 2)],
          i = left,
          j = right;
  while(i <= j) {
    // 使 i 指向第一个(从左至右)大于 pivot 的数
    while(arr[i] < pivot) {
      i++;
    }
    // 使 j 指向第一个(从右至左)小于 pivot 的数
    while(arr[j] > pivot) {
      j--;
    }
    if (i <= j) {
      swap(arr, i, j);
      i++;
      j--;
    }
  }
  return i;
}

function swap(arr, start, end) {
  let temp = arr[start];
  arr[start] = arr[end];
  arr[end] = temp;
}